### PR TITLE
Dashboard & Settings

### PR DESCRIPTION
**Pull request by samisuteria**
WIP

- Update parts of dashboard cells to match current designs
- Added basics settings screen

To do:
- Facebook caching
- API Integration
- Image caching

Simulator Screen Shot - iPhone Xʀ - 2019-09-08 at 02 57 54

Simulator Screen Shot - iPhone Xʀ - 2019-09-08 at 17 41 54

### CODE REVIEW

File: WalkingChallenge/DashboardV2/DashboardDataSource.swift [hunk 1]
```diff
@@ -0,0 +1,88 @@
...
+    onBackground { [weak self] in
+      Facebook.getRealName(for: "me") { (name) in
+        self?.name = name ?? "Could not load"
+        self?.configureCells()
+        completion()
+      }
+    }
+  }
+  
+  func configureCells() {
```
Comment (samisuteria) [inline WalkingChallenge/DashboardV2/DashboardDataSource.swift]: A lot of the boilerplate and refactoring is done so that all screens can use table views with small data sources backing them instead of putting networking/layout/logic into views/view controllers.

File: .swiftlint.yml [hunk 2]
```diff
@@ -1,5 +1,6 @@
 disabled_rules: # rule identifiers to exclude from running
   - todo
   - closure_parameter_position
+  - trailing_whitespace # this is default Xcode behavior
```
Comment (compnerd) [inline .swiftlint.yml]: Please clean up the trailing whitespace instead

File: WalkingChallenge/StyleGuide.swift [hunk 3]
```diff
@@ -51,6 +51,7 @@ struct Style {
 
     static let FoundationGrey = #colorLiteral(red: 0.0784313725, green: 0.0784313725, blue: 0.0784313725, alpha: 1)       // #333333
     static let Silver = #colorLiteral(red: 0.4980392156, green: 0.4980392156, blue: 0.4980392156, alpha: 1)               // #7f7f7f
+    static let Background = #colorLiteral(red: 0.9647058824, green: 0.9725490196, blue: 0.9803921569, alpha: 1)           // #F6F8FA
```
Comment (compnerd) [inline WalkingChallenge/StyleGuide.swift]: Can you add a comment indicating that this is not in the official AKF style guide but is part of the app style only?

File: WalkingChallenge/Context.swift [hunk 4]
```diff
@@ -0,0 +1,36 @@
+/**
+ * Copyright © 2017 Aga Khan Foundation
```
Comment (compnerd) [inline WalkingChallenge/Context.swift]: 2019 please

File: WalkingChallenge/Cells.swift [hunk 5]
```diff
@@ -0,0 +1,79 @@
+/**
+ * Copyright © 2017 Aga Khan Foundation
```
Comment (compnerd) [inline WalkingChallenge/Cells.swift]: 2019 please

File: WalkingChallenge/Cells.swift [hunk 6]
```diff
@@ -0,0 +1,79 @@
...
+ * OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY,
+ * WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR
+ * OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF
+ * ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ **/
+
+import UIKit
+
+protocol CellContext: Context {
+  var cellIdentifier: String { get }
```
Comment (compnerd) [inline WalkingChallenge/Cells.swift]: Why not `identifier`?

File: WalkingChallenge/Cells.swift [hunk 7]
```diff
@@ -0,0 +1,79 @@
...
+
+protocol ConfigurableCell: class {
+  func configure(context: CellContext)
+}
+
+typealias ConfigurableTableViewCell = ConfigurableCell & TableViewCell
+typealias ConfigurableCollectionViewCell = ConfigurableCell & UICollectionViewCell
+
+protocol ListDataSource {
+  var cells: [[CellContext]] { get set }
```
Comment (compnerd) [inline WalkingChallenge/Cells.swift:44]: Why the array of arrays?

File: WalkingChallenge/Cells.swift [hunk 8]
```diff
@@ -0,0 +1,79 @@
...
+  func reload(completion: @escaping GenericBlock)
+  func configureCells()
```
Comment (compnerd) [inline WalkingChallenge/Cells.swift]: Why not `configure`?

File: WalkingChallenge/Support/Blocks.swift [hunk 9]
```diff
@@ -0,0 +1,35 @@
+/**
+ * Copyright © 2017 Aga Khan Foundation
```
Comment (compnerd) [inline WalkingChallenge/Support/Blocks.swift]: 2019 please

File: WalkingChallenge/Support/Blocks.swift [hunk 10]
```diff
@@ -0,0 +1,35 @@
...
+ * OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF
+ * ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ **/
+
+import Foundation
+
+typealias GenericBlock = () -> Void
+typealias PassBlock<T> = (T) -> Void
+typealias ReturnBlock<T> = () -> T
+typealias TransformBlock<A,B> = (A) -> B
```
Comment (compnerd) [inline WalkingChallenge/Support/Blocks.swift]: I'm not really sure I agree that these are useful.

File: WalkingChallenge/Support/UIRefreshControl.swift [hunk 11]
```diff
@@ -0,0 +1,38 @@
+/**
+ * Copyright © 2017 Aga Khan Foundation
```
Comment (compnerd) [inline WalkingChallenge/Support/UIRefreshControl.swift]: 2019 please

File: WalkingChallenge/Support/UITableView.swift [hunk 12]
```diff
@@ -0,0 +1,65 @@
+/**
+ * Copyright © 2017 Aga Khan Foundation
```
Comment (compnerd) [inline WalkingChallenge/Support/UITableView.swift]: 2019 please

File: WalkingChallenge/Support/UITableView.swift [hunk 13]
```diff
@@ -0,0 +1,65 @@
...
+    estimatedRowHeight = WalkingChallenge.Style.Padding.p40
+    rowHeight = UITableView.automaticDimension
+    separatorStyle = .none
+    delegate = viewController
+    dataSource = viewController
+    registerAllCells()
+  }
+  
+  func registerAllCells() {
+
```
Comment (compnerd) [inline WalkingChallenge/Support/UITableView.swift]: Should this raise an error?

File: WalkingChallenge/Support/UITableView.swift [hunk 14]
```diff
@@ -0,0 +1,65 @@
...
+      self.reloadData()
+    }
+  }
+  
+  func dequeueAndConfigureReusableCell(dataSource: TableViewDataSource?, indexPath: IndexPath) -> ConfigurableTableViewCell {
+    guard
+      let cellContext = dataSource?.cell(for: indexPath),
+      let cell = dequeueReusableCell(withIdentifier: cellContext.cellIdentifier, for: indexPath) as? ConfigurableTableViewCell
+    else {
+      print("Trying to dequeue a cell that was not registered")
```
Comment (compnerd) [inline WalkingChallenge/Support/UITableView.swift]: Should this be an error instead?

File: WalkingChallenge/CommonUI/EmptyCell.swift [hunk 15]
```diff
@@ -0,0 +1,46 @@
+/**
+ * Copyright © 2017 Aga Khan Foundation
```
Comment (compnerd) [inline WalkingChallenge/CommonUI/EmptyCell.swift]: 2019

File: WalkingChallenge/CommonUI/EmptyTableViewDataSource.swift [hunk 16]
```diff
@@ -0,0 +1,38 @@
+/**
+ * Copyright © 2017 Aga Khan Foundation
```
Comment (compnerd) [inline WalkingChallenge/CommonUI/EmptyTableViewDataSource.swift]: 2019

File: WalkingChallenge/CommonUI/NavigationController.swift [hunk 17]
```diff
@@ -0,0 +1,41 @@
+/**
+ * Copyright © 2017 Aga Khan Foundation
```
Comment (compnerd) [inline WalkingChallenge/CommonUI/NavigationController.swift]: 2019

File: WalkingChallenge/CommonUI/TableViewCell.swift [hunk 18]
```diff
@@ -0,0 +1,47 @@
+/**
+ * Copyright © 2017 Aga Khan Foundation
```
Comment (compnerd) [inline WalkingChallenge/CommonUI/TableViewCell.swift]: 2019

File: WalkingChallenge/CommonUI/TableViewCell.swift [hunk 19]
```diff
@@ -0,0 +1,47 @@
...
+ * OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY,
+ * WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR
+ * OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF
+ * ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ **/
+
+import UIKit
+
+class TableViewCell: UITableViewCell {
+
```
Comment (compnerd) [inline WalkingChallenge/CommonUI/TableViewCell.swift]: Unnecessary whitespace

File: WalkingChallenge/CommonUI/TableViewController.swift [hunk 20]
```diff
@@ -0,0 +1,109 @@
+/**
+ * Copyright © 2017 Aga Khan Foundation
```
Comment (compnerd) [inline WalkingChallenge/CommonUI/TableViewController.swift]: 2019

File: WalkingChallenge/CommonUI/ViewController.swift [hunk 21]
```diff
@@ -0,0 +1,42 @@
+/**
+ * Copyright © 2017 Aga Khan Foundation
```
Comment (compnerd) [inline WalkingChallenge/CommonUI/ViewController.swift]: 2019

File: WalkingChallenge/Assets.xcassets/BlueBackButton.imageset/Contents.json [hunk 22]
```diff
@@ -0,0 +1,12 @@
+{
+  "images" : [
+    {
+      "idiom" : "universal",
+      "filename" : "BlueBackButtono.pdf"
```
Comment (compnerd) [inline WalkingChallenge/Assets.xcassets/BlueBackButton.imageset/Contents.json]: `BlueBlackButtono.pdf`?

File: WalkingChallenge/StyleGuide.swift [hunk 23]
```diff
@@ -34,24 +34,26 @@ struct Style {
   
   struct Colors {
     // Standard
-    static let black = #colorLiteral(red: 0.2000000000, green: 0.2000000000, blue: 0.2000000000, alpha: 1)      // #333333
-    static let blue = #colorLiteral(red: 0.0509803921, green: 0.4549019607, blue: 0.8901960784, alpha: 1)       // #0d74e3
-    static let green = #colorLiteral(red: 0.0000000000, green: 0.6588235294, blue: 0.3647058823, alpha: 1)      // #00a85d
-    static let grey = #colorLiteral(red: 0.4980392156, green: 0.4980392156, blue: 0.4980392156, alpha: 1)       // #7f7f7f
-    static let red = #colorLiteral(red: 0.8078431372, green: 0.0000000000, blue: 0.3450980392, alpha: 1)        // #ce0058
-    static let white = #colorLiteral(red: 1.0000000000, green: 1.0000000000, blue: 1.0000000000, alpha: 1)      // #ffffff
+    static let black = UIColor(named: "black")!
```
Comment (compnerd) [inline WalkingChallenge/StyleGuide.swift:37]: Please revert this part of the change.  These are colours as per the AKF style guidelines, we do not control them.

File: WalkingChallenge/Support/CALayer.swift [hunk 24]
```diff
@@ -0,0 +1,47 @@
+/**
+ * Copyright © 2017 Aga Khan Foundation
```
Comment (compnerd) [inline WalkingChallenge/Support/CALayer.swift]: 2019

File: WalkingChallenge/Support/UIImage.swift [hunk 25]
```diff
@@ -0,0 +1,46 @@
+/**
+ * Copyright © 2017 Aga Khan Foundation
```
Comment (compnerd) [inline WalkingChallenge/Support/UIImage.swift]: 2019

File: WalkingChallenge/CommonUI/Cells/TableViewCell.swift [hunk 26]
```diff
@@ -42,6 +42,8 @@ class TableViewCell: UITableViewCell {
   }
   
   func commonInit() {
+    backgroundColor = nil
```
Comment (compnerd) [inline WalkingChallenge/CommonUI/Cells/TableViewCell.swift:44]: Why the code changes in the movement?

File: WalkingChallenge/CommonUI/Views/CardView.swift [hunk 27]
```diff
@@ -0,0 +1,40 @@
+/**
+ * Copyright © 2017 Aga Khan Foundation
```
Comment (compnerd) [inline WalkingChallenge/CommonUI/Views/CardView.swift]: 2019

File: WalkingChallenge/CommonUI/Views/CellDisclosureView.swift [hunk 28]
```diff
@@ -0,0 +1,78 @@
+/**
+ * Copyright © 2017 Aga Khan Foundation
```
Comment (compnerd) [inline WalkingChallenge/CommonUI/Views/CellDisclosureView.swift]: 2019

File: WalkingChallenge/CommonUI/Views/View.swift [hunk 29]
```diff
@@ -0,0 +1,46 @@
+/**
+ * Copyright © 2017 Aga Khan Foundation
```
Comment (compnerd) [inline WalkingChallenge/CommonUI/Views/View.swift]: 2019

File: WalkingChallenge/Support/CGFloat.swift [hunk 30]
```diff
@@ -0,0 +1,41 @@
+/**
+ * Copyright © 2017 Aga Khan Foundation
```
Comment (compnerd) [inline WalkingChallenge/Support/CGFloat.swift]: 2019

File: WalkingChallenge/Support/UIBarButtonItem.swift [hunk 31]
```diff
@@ -0,0 +1,37 @@
+/**
+ * Copyright © 2017 Aga Khan Foundation
```
Comment (compnerd) [inline WalkingChallenge/Support/UIBarButtonItem.swift]: 2019

File: WalkingChallenge/Support/UIColor.swift [hunk 32]
```diff
@@ -0,0 +1,79 @@
+/**
+ * Copyright © 2017 Aga Khan Foundation
```
Comment (compnerd) [inline WalkingChallenge/Support/UIColor.swift]: 2019

File: WalkingChallenge/Settings/SettingsDataSource.swift [hunk 33]
```diff
@@ -0,0 +1,48 @@
...
+class SettingsDataSource: TableViewDataSource {
+  var cells: [[CellContext]] = []
+  
+  private var isTeamLead = true
+  
+  func configureCells() {
+    cells = [[
+      SettingsProfileCellContext(
+        image: nil,
+        name: "Sami Suteria",
```
Comment (compnerd) [inline WalkingChallenge/Settings/SettingsDataSource.swift:45]: I think that we have enough of the FB connection to pull the user's real name.

File: WalkingChallenge/Settings/SettingsDataSource.swift [hunk 34]
```diff
@@ -0,0 +1,48 @@
...
+        teamName: "World Walkers",
```
Comment (compnerd) [inline WalkingChallenge/Settings/SettingsDataSource.swift:46]: I think that we want to use some other state that we are pulling this data from, and we can hardcode the strings there.

File: WalkingChallenge/Settings/SettingsViewController.swift [hunk 35]
```diff
@@ -0,0 +1,39 @@
...
+ * ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ **/
+
+import UIKit
+
+class SettingsViewController: TableViewController {
+  override func commonInit() {
+    super.commonInit()
+    
+    title = "Settings"
```
Comment (compnerd) [inline WalkingChallenge/Settings/SettingsViewController.swift]: Move to `Strings`?

File: WalkingChallenge/Settings/Cells/SettingsDisclosureCell.swift [hunk 36]
```diff
@@ -0,0 +1,110 @@
...
+
+struct SettingsDisclosureCellContext: CellContext {
+  let cellIdentifier: String = SettingsDisclosureCell.identifier
+  let title: String
+  let body: String?
+  let value: String?
+  let isDisclosureHidden: Bool
+  let isLastItem: Bool
+  
+  init(title: String, body: String? = nil, value: String? = nil, isDisclosureHidden: Bool = false, isLastItem: Bool = false) {
```
Comment (compnerd) [inline WalkingChallenge/Settings/Cells/SettingsDisclosureCell.swift]: I really wish that Swift would synthesize the constructors

File: WalkingChallenge/Settings/Cells/SettingsDisclosureCell.swift [hunk 37]
```diff
@@ -0,0 +1,110 @@
...
+    contentView.addSubview(disclosureImageView) {
+      $0.top.equalToSuperview().inset(Style.Padding.p16)
+      $0.trailing.equalToSuperview().inset(Style.Padding.p32)
+      $0.height.equalTo(16)
+      $0.width.equalTo(10)
+    }
+    
+    contentView.addSubview(valueLabel) {
+      $0.leading.equalTo(layoutGuide.snp.trailing)
+      $0.trailing.equalTo(disclosureImageView.snp.leading).offset(-Style.Padding.p8)
```
Comment (compnerd) [inline WalkingChallenge/Settings/Cells/SettingsDisclosureCell.swift:98]: s/offset/inset/ and multiply by `-1`?

File: WalkingChallenge/Settings/Cells/SettingsSwitchCell.swift [hunk 38]
```diff
@@ -0,0 +1,125 @@
...
+ * OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY,
+ * WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR
+ * OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF
+ * ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ **/
+
+import UIKit
+
+protocol SettingsSwitchCellDelegate: class {
+  func settingsSwitchToggled(newValue: Bool, context: Context?)
```
Comment (compnerd) [inline WalkingChallenge/Settings/Cells/SettingsSwitchCell.swift]: Why not `toggled`?

File: WalkingChallenge/Settings/Cells/SettingsActionCell.swift [hunk 39]
```diff
@@ -0,0 +1,72 @@
...
+ * OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY,
+ * WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR
+ * OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF
+ * ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ **/
+
+import UIKit
+
+protocol SettingsActionCellDelegate: class {
+  func settingsActionCellTapped(context: Context?)
```
Comment (compnerd) [inline WalkingChallenge/Settings/Cells/SettingsActionCell.swift:33]: Why not `tapped`?

File: WalkingChallenge/Settings/TeamSettingsViewController.swift [hunk 40]
```diff
@@ -0,0 +1,49 @@
...
+import UIKit
+
+class TeamSettingsViewController: TableViewController {
+  override func commonInit() {
+    super.commonInit()
+    
+    title = "Team"
+    dataSource = TeamSettingsDataSource()
+    navigationItem.rightBarButtonItem = UIBarButtonItem(
+      title: "Edit",
```
Comment (compnerd) [inline WalkingChallenge/Settings/TeamSettingsViewController.swift]: Move to `Strings`?

File: WalkingChallenge/Settings/TeamSettingsViewController.swift [hunk 41]
```diff
@@ -0,0 +1,49 @@
...
+ * ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ **/
+
+import UIKit
+
+class TeamSettingsViewController: TableViewController {
+  override func commonInit() {
+    super.commonInit()
+    
+    title = "Team"
```
Comment (compnerd) [inline WalkingChallenge/Settings/TeamSettingsViewController.swift]: Move to `Strings`?

**Review by compnerd — CHANGES_REQUESTED** [review]

File: WalkingChallenge/Assets.xcassets/BlueBackButton.imageset/Contents.json (hunk 22 above)
Comment (samisuteria) [reply WalkingChallenge/Assets.xcassets/BlueBackButton.imageset/Contents.json]: Thats what the layer from the sketch file exported as. Cleaning up name.

File: WalkingChallenge/Cells.swift (hunk 7 above)
Comment (samisuteria) [reply WalkingChallenge/Cells.swift:44]: to support sections of items/rows in collection/tableviews.

File: WalkingChallenge/CommonUI/Cells/TableViewCell.swift (hunk 26 above)
Comment (samisuteria) [reply WalkingChallenge/CommonUI/Cells/TableViewCell.swift:44]: to match cell styling in the sketch files

File: WalkingChallenge/Settings/Cells/SettingsActionCell.swift (hunk 39 above)
Comment (samisuteria) [reply WalkingChallenge/Settings/Cells/SettingsActionCell.swift:33]: Typically a view controller conforms to multiple cell delegates and this helps differentiate which cell's delegate method is being called. Its also Cocoa convention to put the calling call into the delegate method call to prevent collisions.

File: WalkingChallenge/Settings/Cells/SettingsDisclosureCell.swift (hunk 37 above)
Comment (samisuteria) [reply WalkingChallenge/Settings/Cells/SettingsDisclosureCell.swift:98]: inset is typically used for matching edges (trailing to trailing) where as offset is used for mismatched.

File: WalkingChallenge/Settings/Cells/SettingsSwitchCell.swift (hunk 38 above)
Comment (samisuteria) [reply WalkingChallenge/Settings/Cells/SettingsSwitchCell.swift]: :+1:

File: WalkingChallenge/Settings/SettingsDataSource.swift (hunk 33 above)
Comment (samisuteria) [reply WalkingChallenge/Settings/SettingsDataSource.swift:45]: This is just a WIP. I didn't want to pull the name again in this class. I plan on caching the facebook responses in between screen transitions

File: WalkingChallenge/Settings/SettingsDataSource.swift (hunk 34 above)
Comment (samisuteria) [reply WalkingChallenge/Settings/SettingsDataSource.swift:46]: I plan on adding stubs at the same time as tests so we can test datasources are configuring cells correctly

File: WalkingChallenge/StyleGuide.swift (hunk 23 above)
Comment (samisuteria) [reply WalkingChallenge/StyleGuide.swift:37]: The colors are still the same as the AKF style guide - they are just moved to the asset catalog to take advantage of iOS 13's upcoming dark mode. 

Screen Shot 2019-09-13 at 11 26 44 PM

File: WalkingChallenge/Support/UITableView.swift (hunk 14 above)
Comment (samisuteria) [reply WalkingChallenge/Support/UITableView.swift]: I don't feel comfortable throwing an error here until tests are added. A missing registration of a cell will cause the app to crash. Returning an EmptyCell() will just not render the content instead.

Comment (kamranpirwani) [reply WalkingChallenge/Support/UITableView.swift]: I'd recommend an assertion here instead which will crash on debug builds and gracefully degrade in production. Right now, if this case happens, it's gracefully degrading in a dev environment which is masking the issue

File: WalkingChallenge/StyleGuide.swift (hunk 23 above)
Comment (kamranpirwani) [reply WalkingChallenge/StyleGuide.swift:37]: Is putting this in the asset catalog safe to do if trying to support devices < iOS13? I was under the impression you could not use this if not targeting iOS13+

Comment (samisuteria) [reply WalkingChallenge/StyleGuide.swift:37]: Yea its safe to use, devices prior to iOS13 just use "Any Appearance". The screen shot above is from Xcode 10.3 targeting iOS 11 and it works fine.

**Comment by samisuteria** [comment]
Simulator Screen Shot - iPhone Xʀ - 2019-09-17 at 22 33 44
